### ISSUE TITLE
fermionic simple update collapses to exactly zero: #667's fermionic half was never fixed

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#878]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

`fpeps()` returns a state that is **exactly zero**. Reproduced on `main` at `55691da`, at the smallest possible bond dimension.

#667's title is *"Simple update collapses iPEPS state (bosonic → near-classical rank-3; **fermionic → zero norm**)"*. #844 fixed the bosonic half in `ipeps_simple_update.py`. `fermionic_ipeps.py` has its **own** simple-update implementation, and it still contains the defect #667 identified. The issue is closed.

## Measured

D=2, `dt=0.05`, `V=0`, random init, tracking the sweep directly:

| step | `\|A\|` | `max(lam_h)` | `min(lam_h)` | `\|A_phys\|` |
|---|---|---|---|---|
| 0 | 4.2318e+00 | 1.0000e+00 | 1.0000e+00 | 4.2318e+00 |
| 1 | 1.0000e+00 | 1.0000e+00 | 2.9258e-01 | 7.1620e-01 |
| 2 | 1.0000e+00 | 1.0000e+00 | 1.9897e-02 | 7.7407e-01 |
| 3 | 1.0000e+00 | 1.0000e+00 | 2.8988e-03 | 9.3340e-01 |
| 5 | 1.0000e+00 | 1.0000e+00 | 1.0936e-07 | 4.7406e-02 |
| **10** | **0** | **0** | **0** | **0** |

`lam_h` and `lam_v` both end as `[0. 0.]`. `min(lam_h)` falls geometrically — 2.9e-01, 2.0e-02, 2.9e-03, 1.1e-07 — then everything is exactly zero and stays there. The default config is `num_imaginary_steps=200`, so every shipped `fpeps()` run is far past this.

Note `|A|` reads a healthy `1.0` at steps 1–5 because `_normalize_tensor` runs last: the tensor is renormalised on the way down, so a norm check on `A` alone cannot see the collapse coming. Only the bond spectrum shows it.

## Root cause — #667, verbatim

`_fpeps_simple_update_horizontal`, `fermionic_ipeps.py:280-299`:

```python
    U, sigma, Vh, s_full = truncated_svd(...)

    # 6. Normalize lambda
    lam_h_new = sigma / (jnp.max(sigma) + EPS)      # <- sigma stored as lambda
    ...
    # 8. Absorb sqrt(sigma) into the bond axis "r"
    sqrt_sig = jnp.sqrt(sigma + EPS)
    U_final = scale_bond_axis(U_final, "r", sqrt_sig)   # <- AND baked into Gamma
```

`sigma` is stored as the new `lam_h` **and** its square root is absorbed into `Gamma`. On the next step `_absorb_lambdas` (line 223-226) scales `r` by `lam_h` again, so the bond carries `lambda**1.5`. That is exactly the mechanism #667 identified, and #844's fix for it was to leave `Gamma` **bare** — the bosonic code now says so explicitly:

```python
    #    Gamma stays BARE (#667): the shared bond's weight lives in lam_h_new and
    #    is re-absorbed in full by step 1 of the next sweep.  Scaling it in here
    #    as well made the bond carry lambda**1.5, which drove the state to the
    #    product state
```

`_fpeps_simple_update_vertical` (line ~362) has the same pair of steps.

## Three further defects in the same functions

Each is independently a bug; the `lambda**1.5` above is on its own sufficient to explain the collapse, so these are listed for completeness rather than as competing explanations.

1. **The additive epsilon (#748 / #865).** `sigma / (jnp.max(sigma) + EPS)` at `:291` and `:373`. #875 replaced this with a relative guard in `ipeps_simple_update.py` because it *inverts* once `max(sigma)` approaches `EPS = 1e-15`, returning a spectrum whose max is 0.643 rather than 1. Given the trajectory above passes 1e-07 by step 5, this is reached. `jnp.sqrt(sigma + EPS)` at `:298` is a third additive epsilon on the same values.

2. **The shared bond is double-counted in the update.** `_absorb_lambdas` scales **all four** legs including `r`, and both ends of the contracted bond come from that same tensor (`A_left = A_abs.relabel("r", "shared")`, `B_right = A_abs.relabels({"l": "shared", ...})`), so the shared bond receives `lam_h` twice. The bosonic path avoids this deliberately: *"r<-lam_h_far (**NOT l** -- that is the shared bond, already on A.r)"*.

3. **`fpeps()` builds the physical tensor with full `lambda`, not `sqrt(lambda)`.** `fpeps()` at `:553` calls `_absorb_lambdas(A_opt, lam_h, lam_v)`, which applies the full `lam` to every leg. The bosonic `_to_physical_tensor` applies `sqrt(lam)` precisely so that each lattice bond picks up `sqrt(lam) * sqrt(lam) = lam` exactly once. As written, each bond of the returned state carries `lam**2`.

## Reproduction

```python
import jax, jax.numpy as jnp
from tenax.algorithms.fermionic_ipeps import (
    FPEPSConfig, _fpeps_simple_update_horizontal, _fpeps_simple_update_vertical,
    _initialize_fpeps, _trotter_gate, spinless_fermion_gate,
)
jax.config.update("jax_enable_x64", True)

cfg = FPEPSConfig(D=2, dt=0.05, V=0.0)
gate = _trotter_gate(spinless_fermion_gate(cfg), cfg.dt)
A = _initialize_fpeps(cfg, jax.random.PRNGKey(0))
lam_h = lam_v = jnp.ones(cfg.D)

for step in range(11):
    A, lam_h = _fpeps_simple_update_horizontal(A, gate, lam_h, lam_v, cfg.D)
    A, lam_v = _fpeps_simple_update_vertical(A, gate, lam_h, lam_v, cfg.D)
    print(step, float(A.norm()), float(jnp.min(lam_h)))
# -> 0.0 0.0 by step 10
```

## Why it survived

The same reason #667 did: **no test asserts a fermionic simple-update energy or bond spectrum.** A zero state is finite and well-formed, so `isfinite` passes on it, and `_normalize_tensor` keeps `|A|` at 1.0 until the moment it is exactly 0. #667 was closed on the bosonic fix without the fermionic half being re-checked.

This is also the "fix landed on some of N copies" pattern that produced #828, #829 and #842 — and #875 is now another instance of it, having fixed the additive epsilon in one of the two modules that had it.

## Suggested fix

Not "port #844's diff". The three defects above mean the fermionic update needs to be re-derived against the bosonic one, with a guard that asserts a *spectrum*, not a norm — `min(lam) > 0` and a converged energy, since `|A|` is actively misleading here.

Note the fermionic path genuinely needs `base_charges` (#558/#559/#563): it is a **1-site** ansatz where `A.l` and `A.r` are the same bond, so the layout constraint #865 removed from the bosonic path must stay here. Any fix has to keep that.

Refs #667, #844, #748, #865, #875, #558, #559, #563.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

**Update: five of the six defects here are fixed, and the residual one is not what it looks like — it is seed-dependent at every bond dimension, not a D≥3 bug.**

## Fixed

Four Vidal-bookkeeping defects, each pinned by a test that failed first:

- `Gamma` no longer absorbs `sqrt(sigma)` while the same `sigma` is stored as the new lambda (the `lambda**1.5` bond — #667's root cause)
- the normalisation is relative, via the **shared** `_normalise_lambda` rather than a second inline copy, so it cannot drift from the bosonic one again (#748/#865)
- the shared bond carries `lambda` once, not twice
- `_to_physical_fpeps_tensor` applies `sqrt(lambda)`, so each lattice bond picks the weight up exactly once (it was carrying `lambda**2`)

And the fifth, structural one: `Vh` was computed at both SVD sites and **discarded**, so on a 1-site ansatz — where `A` is *both* ends of every bond — `A` received the left/top half of every gate and never the right/bottom. That is the asymmetry #667 identified. Confirmed by a `dt` sweep: at equal imaginary time, `dt` = 0.05 / 0.01 / 0.005 **all** gave `lam_h = [1, 0]`, i.e. dt-independent, so a fixed-point property rather than Trotter error.

fPEPS therefore moves to the 2-site checkerboard end to end (shared sweep from #877, `ctm_split_tensor_2site` for the environment). The pair is not cosmetic — the t-V ground state is a charge-density wave at finite `V`, which no single tensor can represent, and the measured sublattice gap after 200 steps is 2.7e-01 at D=2/V=0, 5.0e-01 at D=3/V=0 and 1.4e+00 at D=2/V=4.

**Result: D=2 gives `lam_h = [1, 0.678]` after 100 steps, against `[1, 0]` before.**

## The residual failure is seed-dependent, not D=3

Seeds 0–4, 600 steps, `dt=0.05`, counting seeds whose bond spectrum stays alive:

| D | seeds alive | dead seeds |
|---|---|---|
| 2 | 4/5 | 1 |
| 3 | **2/5** | 0, 1, 3 |
| 4 | 4/5 | 1 |
| 6 | 4/5 | 0 |

Every bond dimension has both surviving and dying seeds. D=3 is the worst survival rate, not a distinct failure mode.

**Seed 0 — which I used for every measurement in this investigation — dies at D=3 and D=6 and lives at D=2 and D=4.** That single accident produced the entire "D=2 works, D=3 breaks, D=4 works" pattern.

## Four hypotheses refuted, all by control arms

1. **The layout pin (`_truncation_base_charges`) is misplaced.** It is called *only* from the 2-site routines, while the 1-site fermionic path pins independently with its own `base_charges=virt_charges` — so its docstring justifies itself by citing a path that does not call it. That reading is correct about the code and wrong about the consequence: **unpinned collapses at every D**, including the D=2 and D=4 that currently work. Worth recording precisely because at D=3/200 steps the unpinned arm returns `[1, 0.874, 0.041]` and looks like a clean fix. Stopping there would have shipped a regression.
2. **Odd D.** Refuted — D=5 `[0,1,0,1,0]` survives.
3. **Unbalanced virtual charges.** Refuted — balanced D=6 dies, unbalanced D=5 lives.
4. **"D=3 is special."** Refuted by the seed sweep.

## What this means

This is the same basin behaviour as [#869 on the bosonic path](https://github.com/tenax-lab/tenax/issues/869#issuecomment-5295625718) (0/3 seeds diverging at D=2, 1/3 at D=3, 3/3 at D=4) — now independently confirmed on the fermionic path, in an implementation that shares nothing with the bosonic one except the stored-lambda state model.

That is the argument for fixing it by **not storing lambda during imaginary-time evolution** rather than by anything local: a non-unitary gate invalidates the stored spectra on every *other* bond, and with the gauge re-derived from the tensors each step there is no opening transient to land on the wrong side of. Design: `docs/superpowers/specs/2026-08-14-simple-update-rewrite-design.md`.

## Still open

- The basin failure above. Guard kept as a **strict xfail pinned to seed 0** — a known-dying seed, so it regression-guards the worst case. It should not be "fixed" by choosing a luckier seed.
- The 2-site energy reads ≈ −5e−5 on a healthy spectrum and full-rank corner. That cannot currently be distinguished from a pre-existing defect, because no fermionic CTM energy path is asserted against a reference anywhere — filed as **#879**.

Work is on `fix/878-fermionic-2site`. No PR yet: this issue is not closable while the basin failure and #879 stand.

Refs #667, #748, #865, #869, #877, #879, #558, #559, #563.
